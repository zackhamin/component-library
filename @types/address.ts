import { CountryCode } from './countryCode';

export interface Address {
	addressDetails: {
		line1: string;
		line2: string | null;
		city: string;
		county: string | null;
		country: CountryCode;
		postCode: string;
		isDefaultAddress: boolean;
		addressType: string | null;
		addressNickName: string | null;
	};
	addressContactDetails: {
		firstName: string;
		lastName: string;
		mobileNumber: string | null;
	};
	addressMetadata: {
		addressId: string;
		datetimeCreated: Date | string;
		datetimeUpdated: Date | string;
		addressType: string | null;
	};
}
