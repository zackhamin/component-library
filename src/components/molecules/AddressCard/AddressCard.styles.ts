import { colors } from 'Styles';
import styled from 'styled-components';

const StyledAddressCard = styled.div`
	border: 1px solid ${colors.grey200};
	border-radius: 4px;
	padding: 1.6rem;
	display: flex;
	flex-direction: column;
	justify-content: space-between;
	gap: 1.6rem;
`;

const StyledButtonList = styled.div`
	display: flex;
	gap: 1.6rem;
`;

export { StyledAddressCard, StyledButtonList };
