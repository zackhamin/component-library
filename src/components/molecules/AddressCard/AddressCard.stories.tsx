import { mockAddress } from 'MockData/address';
import { AddressCard, AddressCardProps } from './AddressCard';

export default {
	title: 'molecules/AddressCard',
	component: AddressCard,
};

export const AddressCardStory = (args: AddressCardProps) => (
	<AddressCard {...args} />
);

AddressCardStory.storyName = 'AddressCard';
AddressCardStory.args = {
	address: mockAddress,
};
AddressCardStory.argTypes = {
	editHandler: { action: 'editing' },
	deleteHandler: { action: 'deleted' },
	setDefaultHandler: { action: 'default set' },
};
