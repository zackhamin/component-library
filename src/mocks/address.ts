import { Address } from '@Types/address';

export const mockAddress: Address = {
	addressDetails: {
		line1: '11 High Road',
		line2: null,
		city: 'London',
		county: null,
		country: 'GB',
		postCode: 'SW1 2AB',
		isDefaultAddress: false,
		addressType: null,
		addressNickName: null,
	},
	addressContactDetails: {
		firstName: 'John',
		lastName: 'Doe',
		mobileNumber: '07712345678',
	},
	addressMetadata: {
		addressId: '8d063d44-1e27-4a38-9602-552730b90552',
		datetimeCreated: '2022-01-01 00:00:00 UTC',
		datetimeUpdated: '2022-01-01 00:00:00 UTC',
		addressType: null,
	},
};
