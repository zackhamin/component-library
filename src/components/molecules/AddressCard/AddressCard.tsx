import { Address as AddressType } from '@Types/address';
import { StyledAddressCard, StyledButtonList } from './AddressCard.styles';

import { Address } from 'Atoms/Address';
import { Button } from 'Atoms/Button';
import { Heading } from 'Atoms/Heading';

export interface AddressCardProps {
	address: AddressType;
	editHandler: () => void;
	deleteHandler: () => void;
	setDefaultHandler: () => void;
	i18n: {
		strings: {
			defaultAddress: string;
			edit: string;
			remove: string;
			setDefault: string;
		};
	};
}

const AddressCard = ({
	address,
	editHandler,
	deleteHandler,
	setDefaultHandler,
	i18n = {
		strings: { defaultAddress: '', edit: '', remove: '', setDefault: '' },
	},
}: AddressCardProps) => {
	const { isDefaultAddress } = address.addressDetails;

	return (
		<StyledAddressCard>
			<div>
				{isDefaultAddress && (
					<Heading level={3} size="1.8rem">
						{i18n.strings.defaultAddress}
					</Heading>
				)}
				<Address
					address={address}
					options={{
						multipleLines: true,
						showName: true,
					}}
				/>
			</div>

			<StyledButtonList>
				<Button size="small" style="text" onClick={editHandler}>
					{i18n.strings.edit}
				</Button>
				<Button
					size="small"
					style="text"
					onClick={deleteHandler}
					disabled={isDefaultAddress}
				>
					{i18n.strings.remove}
				</Button>
				{!isDefaultAddress && (
					<Button size="small" style="text" onClick={setDefaultHandler}>
						{i18n.strings.setDefault}
					</Button>
				)}
			</StyledButtonList>
		</StyledAddressCard>
	);
};

export { AddressCard };
