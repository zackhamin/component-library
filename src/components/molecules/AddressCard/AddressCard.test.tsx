import { AddressCard } from '../AddressCard';
import { mockAddress } from 'MockData/address';
import { fireEvent, render } from '@testing-library/react';

export const mocki18n = {
	strings: {
		defaultAddress: 'general.default_address',
		edit: 'general.edit',
		remove: 'general.remove',
		setDefault: 'general.set_default',
	},
};

describe('AddressCard', () => {
	it('renders correctly', () => {
		const { container } = render(
			<AddressCard
				address={mockAddress}
				editHandler={jest.fn}
				deleteHandler={jest.fn}
				setDefaultHandler={jest.fn}
				i18n={mocki18n}
			/>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('renders default address correctly', () => {
		const { getByText } = render(
			<AddressCard
				address={{
					...mockAddress,
					addressDetails: {
						...mockAddress.addressDetails,
						isDefaultAddress: true,
					},
				}}
				editHandler={jest.fn}
				deleteHandler={jest.fn}
				setDefaultHandler={jest.fn}
				i18n={mocki18n}
			/>
		);
		expect(getByText(/general.default_address/i)).toBeInTheDocument();
	});

	it('calls event handlers', () => {
		const editHandler = jest.fn();
		const removeHandler = jest.fn();
		const setDefaultHandler = jest.fn();

		const { getByText } = render(
			<AddressCard
				address={mockAddress}
				editHandler={editHandler}
				deleteHandler={removeHandler}
				setDefaultHandler={setDefaultHandler}
				i18n={mocki18n}
			/>
		);

		const editButton = getByText(/general.edit/i);
		const removeButton = getByText(/general.remove/i);
		const setDefaultButton = getByText(/general.set_default/i);

		fireEvent.click(editButton);
		expect(editHandler).toHaveBeenCalledTimes(1);

		fireEvent.click(removeButton);
		expect(removeHandler).toHaveBeenCalledTimes(1);

		fireEvent.click(setDefaultButton);
		expect(setDefaultHandler).toHaveBeenCalledTimes(1);
	});
});
