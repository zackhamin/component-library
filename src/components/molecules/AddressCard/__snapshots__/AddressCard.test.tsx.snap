// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressCard renders correctly 1`] = `
.c0 {
  border: 1px solid #F6F6F6;
  border-radius: 4px;
  padding: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: 1.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1.6rem;
}

.c2 {
  background: #1F51FF;
  border-radius: 0.8rem;
  border: 2px solid #EEEEEF;
  color: #ffffff;
  cursor: pointer;
  font-weight: 700;
  min-width: 18rem;
  padding: 1.8rem 3.2rem;
  text-shadow: 0px 1px 1px rgba(0,0,0,0.25);
  -webkit-transition: 0.2s;
  transition: 0.2s;
  white-space: nowrap;
  line-height: 1;
  font-size: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2:hover,
.c2:hover:not([disabled]) {
  background: #0000FF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2.secondary {
  background: #ffffff;
  border: 2px solid #585858;
  border-radius: 6px;
  color: #585858;
  text-shadow: none;
}

.c2.secondary:hover {
  background: #ffffff !important;
  border: 2px solid #000000;
  color: #000000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2.icon {
  background: none;
  border: none;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-width: 0;
  text-shadow: none;
}

.c2.icon:hover {
  background: none;
}

.c2.icon:focus {
  outline-color: #00594C;
}

.c2.btn-medium {
  font-size: 1.6rem;
  min-width: 16rem;
  padding: 1.2rem 2.4rem;
}

.c2.btn-small {
  font-size: 1.4rem;
  min-width: 15rem;
  padding: 0.8rem 2.4rem;
}

.c2.block {
  display: block;
}

.c2[disabled] {
  cursor: not-allowed !important;
  opacity: 0.5;
}

.c2.text {
  background: transparent;
  border: 0;
  border-radius: 0;
  border-bottom: 1px solid transparent;
  color: #808080;
  cursor: pointer;
  font-size: 1.6rem;
  font-weight: normal;
  min-width: auto;
  padding: 0;
  text-shadow: none;
}

.c2.text:hover {
  background: transparent !important;
  border-bottom: 1px solid #585858;
  color: #000000;
}

.c2.text.btn-small {
  font-size: 1.4rem;
}

.c2.text.btn-medium {
  font-size: 1.6rem;
}

.c2.text.btn-large {
  font-size: 1.8rem;
}

.c2.block {
  width: 100%;
}

<div
  class="c0"
>
  <div>
    <p>
      John
       
      Doe
      <br />
      11 High Road
      <br />
      London
      <br />
      SW1 2AB
      <br />
      GB
    </p>
  </div>
  <div
    class="c1"
  >
    <button
      class="c2 text btn-small"
    >
      general.edit
    </button>
    <button
      class="c2 text btn-small"
    >
      general.remove
    </button>
    <button
      class="c2 text btn-small"
    >
      general.set_default
    </button>
  </div>
</div>
`;
